mycroft/skills/intent_service.py

```diff
@@ -296,14 +321,30 @@ def handle_utterance(self, message):
                     match = match_func(combined, lang, message)
                     if match:
                         break
+
             if match:
                 if match.skill_id:
-                    self.add_active_skill(match.skill_id)
-                    # If the service didn't report back the skill_id it
-                    # takes on the responsibility of making the skill "active"
+                    if self.skill_categories[match.skill_id] == 'undefined':
+                        """
+                        this has not been tested and not sure if this is 
+                        what we want. this will cause all old style skills 
+                        to send out a common play stop message. this is not
+                        current behavior. 
+                        """
+                        # stop last active common play skill
+                        msg_data = {"stop_type":"latest"}
+                        self.bus.emit(message.forward('play:stop', msg_data))
```

Comment:
This seems really weird, can you elaborate of what the intention is to stop any playing media for "undefined" skills. I don't have access to the Jira ticket(s) this PR refers to.

Perhaps it's a good idea to break this out into a named method with a docstring explaining what it does and how it ties into the rest of the system.

Reply:
Ah, good catch Ake. So this is an interesting change which was not present in the previous PR similar to this new one. I believe it was called SkillStates. This change is a direct result of enabling barge-in which was necessary to be able to stop long playing dialog skills like wiki and the definition of some basic system level interactions.

When looking for UI guidance on how skills should interact it was stated that if a skill is playing (say for example a music skill) and another skill gets launched (say while the music is playing I barge-in and say 'tell me about the Beatles') then the old skill should be stopped (music) and the new skill should take over (wiki). It should also work the other way around. If wiki were playing and I said 'play louie louie' the music skill activation should cause the wiki skill to stop. However, if an alarm goes off while the music is playing and I say stop, the alarm should stop but not the music skill. Likewise, while I am playing music, if I say 'set the volume to 50%' that should also, not stop the music skill. The ability to handle different more complex scenarios like this obviously led to the skill category I added in this design, however, I still needed to make some trade-offs. 

For example, the code in question works if the underlying skill is a newer style skill like a common play skill or a statefull skill, however, if the underlying skill is an old style skill (normal skill as they exist today) then they will work as always which is probably not what we will want so that problem still has to be solved. I felt even though this would change default behavior it was halfway to the ultimate solution. 

This is a contentious change and unlike anything else in this PR could be considered a change in existing behavior. I would not be adverse to simply commenting out these two lines for now and leaving the default skill behavior unchanged.

Reply:
can you make this behavior configurable? a `"common_play_barge_in": true` in the .conf file maybe? 
we could either add a whole section for `"common_play"` assuming more config options will be needed in the future, or a simple boolean under the "skills" section

Reply:
This is an interesting 'ask'. So barge-in is currently controlled by a value in the default config file 

"mute_during_output" : true,

Gez has requested in a different comment on this PR to remove the change (I changed it to false) from here and put it into the system conf file. Do you think I should use this value in the decision making process or maybe even should it be configurable at a skill level? Be interested in hearing Gez's thoughts on this as well.

Reply:
here is the scenario im thinking about that makes me think this should be optional in the first place (the mic quality is also a consideration)

- music is playing
- microphone is good, hears wakeword, lowers volume, receives utterance, restores volume
- music is playing like nothing happened
- my intent was a simple non intrusive command

the concern is that this new stop signal makes music stop completely a lot of the time

Reply:
Ah, so yes, you are correct and this was a conscience decision though I could be talked out of it but let me explain. 

The ask from our UI department was that if you were playing a song and I said 'tell me about bla' that the song would stop and the new skill would take over. This was not the desired behavior, however, in the case of a system type event like an alarm or a request to increase the volume. In this case they should not terminate the underlying active skill. 

So, this bit of code does just that. If you look at the alarm and/or the volume skill you will see they set their category to 'system' which basically means they will not stop an underlying skill, but older skills like wiki or ddg will indeed stop the underlying skill. It was a tough decision because this really only solves half the problem. When an underlying skill is an old style skill it will behave as they do now and not stop when another skill gains control of the output channel. I would like to solve this also but have not decided how to do that yet. 

So do you think I should remove these two lines of code until the entire (both halves) problem has been solved or do you think I should leave this here and allow skills to change their category to deal with this. Keep in mind only category = system is considered currently but that could be easily changed.

Reply:
And let me add if you check out the new version of the volume skill in conjunction with this code it will behave as you mention. It will change the volume and resume the media play. Similarly the alarm skill will respond to stop and not take out any underlying skill; new style or old style. It will just stop the alarm.